File: movieland-backend/src/main/java/com/phil/movieland/rest/tasks/TaskService.java
package com.phil.movieland.rest.tasks;

import com.phil.movieland.rest.controller.StatisticsController;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

/**
 * Service to manage asnyc REST Tasks in ThreadPool
 */
@Service
public class TaskService {

    private Logger log = LoggerFactory.getLogger(TaskService.class);

    private ExecutorService executor;
    private List<RunnableWithProgress> taskList = new ArrayList<>();
    private List<Integer> finishedTaskIds = new ArrayList<>();

    public TaskService() {
        this.executor = Executors.newFixedThreadPool(5);
    }

    /**
     * Execute task in ThreadPool and return taskId
     */
    public int execute(RunnableWithProgress runnable) {
        log.info("Executing Task '{}'", runnable.getMessage());
        executor.execute(runnable);
        int taskId;
        if (finishedTaskIds.size() > 0) {
            taskId = finishedTaskIds.remove(0);
            taskList.set(taskId, runnable);
        } else {
            taskList.add(runnable);
            taskId = taskList.size() - 1;
        }
        return taskId;
    }

    /**
     * Return current Progress of Task
     */
    public TaskProgress getTaskProgress(int taskId) {
        if (taskId >= taskList.size()) {
            return null;
        }
        RunnableWithProgress task = taskList.get(taskId);
        if (task == null) {
            return null;
        }
        return new TaskProgress(task.getProgressMax(), task.getProgress(), taskId, task.getMessage());
    }

    public void removeTask(int taskId) {
        log.info("Removing Task '{}'", taskId);
        taskList.set(taskId, null);
        finishedTaskIds.add(taskId);
    }
}

File: movieland-backend/rest/service/task_service.py
from concurrent.futures import ThreadPoolExecutor
from datetime import datetime, timedelta
from typing import List, Any, Dict

from marshmallow import fields

from db.database import db, ma
from db.model import MovieShow, Reservation, Seat
from logger import get_logger

log = get_logger()


class TaskProgress:

    def __init__(self, progress_max: int, progress: int, task_id: int, message: str):
        self.progress_max = progress_max
        self.progress = progress
        self.task_id = task_id
        self.message = message


class TaskProgressSchema(ma.Schema):
    class Meta:
        model = TaskProgress

    progressMax = fields.Integer(attribute='progress_max')
    progress = fields.Integer()
    taskId = fields.Integer(attribute='task_id')
    message = fields.String()


task_progress_schema = TaskProgressSchema()


class RunnableWithProgress:
    progress_max: int = 100
    progress: int = 0
    message: str = None
    runnable = None
    args: Dict[str,Any]


class TasksService:
    task_list: List[RunnableWithProgress] = []
    finished_task_ids: List[int] = []
    executor: ThreadPoolExecutor = ThreadPoolExecutor(max_workers=3)

    def execute(self, runnable: RunnableWithProgress) -> int:
        log.info(f"Executing Task '{runnable.message}'")
        task_id = None

        self.executor.submit(runnable.runnable, **runnable.args)

        if len(self.finished_task_ids) > 0:
            task_id = self.finished_task_ids.pop(len(self.finished_task_ids) - 1)
            self.task_list[task_id] = runnable
        else:
            self.task_list.append(runnable)
            task_id = len(self.task_list) - 1
        return task_id

    def get_task_progress(self, task_id: int) -> TaskProgress:
        if task_id > len(self.task_list):
            return None
        task = self.task_list[task_id]
        if task is None:
            return None
        return TaskProgress(task.progress_max, task.progress, task_id, task.message)
